### ISSUE TITLE
Search - App navigates to Inbox page when submitting an expense via QAB

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#50157]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.44-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+shsb22cp111@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

**Issue found when executing PR** https://github.com/Expensify/App/pull/50028
## Action Performed:
1. Login to any account
2. Switch to the Search page tab
3. Click the global create button (FAB) > Submit expense > Manual > Enter amount > Choose recipient and submit
4. Notice that you remain on the search page
5. Click on the FAB button again > Quick Action Button > Enter amount and submit

## Expected Result:
The user remains on the search page and the newly added expense appears in the table and the row is briefly highlighted

## Actual Result:
The user is navigated to inbox page, there is an inconsistency when submitting an expense via FAB and QAB

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/8332af02-fa13-4c54-8496-908a3e9b7092

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lanitochka17** [comment]
@slafortune  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave-control

**Comment by trjExpensify** [comment]
Yep, like the create options, we'd want someone using the QAB to remain on the search page if they created from the search page as well. 

@mananjadhav can you take this?

**Comment by mananjadhav** [comment]
Will take this up.

**Comment by mananjadhav** [comment]
I have a solution that works. This definitely seems to be related to Navigation so I was verifying if it would break something. I also feel we should get an opinion from the expert agency as it relates to Navigation.

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

- When we click on the "Quick action" from search or other pages on FAB, it navigates to "Inbox" page.

### What is the root cause of that problem?

- We're calling `IOU.startMoneyRequest` with parameters `IOU.startMoneyRequest(CONST.IOU.TYPE.SUBMIT, quickActionReportID, CONST.IOU.REQUEST_TYPE.MANUAL, true)`.
- When this is called, the navigation path doesn't directly called the `/create/submit/start/{reportId}/manual`. Instead this calls `/create/submit/{reportId}?initial=true&screen=scan&params=object&path=/create/submit/start/...`. I think this is related to the `linkTo` implementation.


### What changes do you think we should make in order to solve the problem?
- For all quick action cases in [FloatingActionButtonAndPopover.tsx](https://github.com/Expensify/App/blob/main/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.tsx), we can get rid of the third parameter `requestType` and set it as `undefined`.

https://github.com/Expensify/App/blob/b027baaf086c4a8f57bec48dd1cbbfb3c317726b/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.tsx#L249-L282

The reason I suggested this is because we the function `startMoneyRequest` has four parameters `(iouType: ValueOf<typeof CONST.IOU.TYPE>, reportID: string, requestType?: IOURequestType, skipConfirmation = false)` but across the application we never use the `requestType` argument.

Here's a quick video of the implementation.

https://github.com/user-attachments/assets/845e6ced-77c4-460b-9b18-917b9474c19f


### What alternative solutions did you explore? (Optional)

- Fix this at the navigation level.

**Comment by c3024** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
BottomTab and CentralPane change when we open RHP with QAB in Search Central Pane.
### What is the root cause of that problem?
When we start the flow with QAB the report in the URL param is a valid report and it finds a matching route for RHP route here.
https://github.com/Expensify/App/blob/66aefc37f0d38ea2a2f2246cb037db26f3686cba/src/libs/Navigation/linkingConfig/getAdaptedStateFromPath.ts#L162-L164
and this is pushed to routes with `metaInfo` as `true`.
Since these are both true, we find if there are diff actions that need to be dispatched here and we dispatch them.
https://github.com/Expensify/App/blob/66aefc37f0d38ea2a2f2246cb037db26f3686cba/src/libs/Navigation/linkTo/index.ts#L136-L143
In this case, the bottom tab navigators and central panes are different (Search_Bottom_Tab/Home and Search_Central_Pane and Report respectively). So, these diffs are dispatched which cause the QAB report and Home to be added for the Central Pane and Bottom Tab respectively.
### What changes do you think we should make in order to solve the problem?
When on Search_Central_Pane I have not found any cases where we need to change the bottom tab or central pane when opening RHP. So we can dispatch the diffs only when when do not open the RHP when on Search Central Pane.
So, we can wrap this in a check
https://github.com/Expensify/App/blob/66aefc37f0d38ea2a2f2246cb037db26f3686cba/src/libs/Navigation/linkTo/index.ts#L135-L143
```js
 if (!(isSideModalNavigator(action.payload.name) && lastRoute?.name === "Search_Central_Pane")) {
            const {adaptedState, metainfo} = getAdaptedStateFromPath(path, linkingConfig.config);
            if (adaptedState && (metainfo.isCentralPaneAndBottomTabMandatory || metainfo.isFullScreenNavigatorMandatory)) {
                const diff = getPartialStateDiff(rootState, adaptedState as State<RootStackParamList>, metainfo);
                const diffActions = getActionsFromPartialDiff(diff);
                for (const diffAction of diffActions) {
                    root.dispatch(diffAction);
                }
            }
}
```
### What alternative solutions did you explore? (Optional)

**Comment by c3024** [comment]
I can work on this as contributor and @mananjadhav as C+ if this solution above looks good.

**Comment by mananjadhav** [comment]
I was working on a solution to update linkTo. Let me review your proposal.

**Comment by mananjadhav** [comment]
Yeah I think @c3024's [solution](https://github.com/Expensify/App/issues/50157#issuecomment-2393006987) works. 

> where we need to change the bottom tab or central pane when opening RHP. 

Just need to clarify this.

**Comment by c3024** [comment]
My bad! My solution is not good enough. The issue can happen when we click on QAB from other pages like the settings page with related centre panes.

I think we should keep a global value for tracking if we clicked on QAB and just ignore adding the report here

https://github.com/Expensify/App/blob/66aefc37f0d38ea2a2f2246cb037db26f3686cba/src/libs/Navigation/linkingConfig/getAdaptedStateFromPath.ts#L162-L164

even if it matches with an existing report like [this](https://github.com/c3024/App/tree/qabvalue).

**Comment by trjExpensify** [comment]
How did we approach this for the create flow when actioned from the search page, wouldn't we follow that?

**Comment by c3024** [comment]
Here is the corrected text:

We generate the navigation state required based on the URL. When we use QAB, we have a valid report ID in the URL. In normal cases (that is, other than QAB), we do not have a valid report ID in the URL. The report is chosen on the participants' page.

When we deeplink into an RHP flow and the URL has an existing report ID, we show the report as an underlay when we navigate to it. Therefore, we add this report screen to the navigation state when the URL has an existing report ID.

The URL is the same for the QAB flow as well, so our navigation code pushes the existing report to the central pane, and this navigation to a different bottom navigator (HOME) and central pane (REPORT) occurs. When we click on the normal 'Submit Expense' flow, there is no valid report ID in the URL, and no report route is pushed into the state. This is peculiar to QAB alone due to the initial URL having a valid report ID.

**Comment by c3024** [comment]
The bug occurs in other pages also, not just in the Search Page.


https://github.com/user-attachments/assets/1f9658d8-1e3e-48ee-8b0d-5dbb9cc55b9f

**Comment by mananjadhav** [comment]
> How did we approach this for the create flow when actioned from the search page, wouldn't we follow that?

@c3024 is right. I think it's just how we've implemented @trjExpensify. Hence we have both the options. Either we go ahead an update the Navigation which @c3024 has recommended or what you asked and that's what my proposal does.

**Comment by trjExpensify** [comment]
>I think it's just how we've implemented @trjExpensify. 

Sorry to be clear, this "stay on the search page if you `Submit expense` from the search page" feature is a relatively new feature in the last couple of weeks. So I was wondering how that was implemented, and why we can't do the same for the QAB. CC: @ikevin127 who might have some context on this to chime in as well.

**Comment by trjExpensify** [comment]
Whatcha' think @mananjadhav?

**Comment by mananjadhav** [comment]
I think we need to decide whether we want to update the navigation for this. I feel we should do that. 

We have both the proposals and I think we should ask an internal engineer to decide this.

🎀 👀 🎀 C+ reviewed (adding this to assign an internal engineer). We can decide between @c3024 or me to implement + review the PR.

**Comment by francoisl** [comment]
I'm leaning @mananjadhav's solution as it simplifies the code without adding a new global tracking value, but one clarification about this please: 

> The reason I suggested this is because we the function startMoneyRequest has four parameters (iouType: ValueOf<typeof CONST.IOU.TYPE>, reportID: string, requestType?: IOURequestType, skipConfirmation = false) but across the application we never use the requestType argument.

Are you suggesting also removing [this `switch` block](https://github.com/Expensify/App/blob/69691038cd6e024e14ca42dfe8ca59387e0226f9/src/libs/actions/IOU.ts#L388-L396) and just keeping the default `Navigation.navigate(..)` then? Would the QAB still work if your quick action is Track Distance or Scan?

**Comment by c3024** [comment]
Instead of keeping a global variable we can pass a param to the route like `?isQuickAction=true` and use that to check if we need to add a central report pane or not like [this](https://github.com/c3024/App/tree/quickaction-as-a-param).

In case if we stumble into something difficult with the simplification of code as specified above, we can try this route param approach instead of using a global variable.

**Comment by mananjadhav** [comment]
@francoisl Yes it works even if the quick action is `Track Distance` or `Scan`.  My proposal was to pass undefined as the param. I tried removing the switch and it worked fine too. My concern with switch case removal is whether it was added for deeplinking. But as I said I checked all the method calls and I didn't find a single instance where we're setting the third and fourth param except for QAB.

**Comment by francoisl** [comment]
Ok sounds good, let's go with your solution then @mananjadhav.

Not sure if we need to assign another C+ though in this case, for the review? 🤔

**Comment by mananjadhav** [comment]
I’ll implement tn

**Comment by mananjadhav** [comment]
I’ll implement the PR, but we can assign @c3024 if we want a C+ review.

**Comment by mananjadhav** [comment]
Ohh I didn't see the emoji reactions. I'll get started on the PR and finish by tomorrow.

**Comment by trjExpensify** [comment]
How you looking on a PR for this?

**Comment by mananjadhav** [comment]
I am out today, but I'll finish this by Friday.

**Comment by francoisl** [comment]
Yes I think the current expectation is that we fix those when we run into them – see https://expensify.slack.com/archives/C01GTK53T8Q/p1726169039084589

**Comment by mananjadhav** [comment]
Thanks, I'll try to cover these and raise a PR by weekend then.

**Comment by trjExpensify** [comment]
@mananjadhav mind providing an update here? Thanks!

**Comment by mananjadhav** [comment]
I posted one yesterday on the PR itself [here](https://github.com/Expensify/App/pull/51057#issuecomment-2426021385). I'll have it raised by tomorrow.

**Comment by mananjadhav** [comment]
I am left with one last withOnyx update and I need to modify the selector.

**Comment by trjExpensify** [comment]
Checklist time, @mananjadhav!

**Comment by mananjadhav** [comment]
I thought @c3024 would take care of it as the reviewer. 

I checked the code and I don't think we have an offending PR for this one. But I do feel we should add a regression test step for this.

**Regression Test Proposal**

1. Login to any account
2. Switch to the Search page tab
3. Click the global create button (FAB) > Submit expense > Manual > Enter amount > Choose recipient and submit.
4. Notice that you remain on the search page.
5. Click on the FAB button again > Quick Action Button > Enter amount and submit.
6. The user remains on the search page and the newly added expense appears in the table.

wdyt?

**Comment by c3024** [comment]
>I thought @c3024 would take care of it as the reviewer.

Yea, I missed that this was due for payment yesterday.

Thanks for handling it!

**Comment by trjExpensify** [comment]
Paid, closing!

**Comment by mananjadhav** [comment]
> Yea, I missed that this was due for payment yesterday.

No worries at all bro!

thanks @trjExpensify

**Comment by garrettmknight** [comment]
$250 approved for @mananjadhav
